go/mysql/flavor_mysql.go

```diff
@@ -345,17 +345,40 @@ GROUP BY t.table_name, t.table_type, t.create_time, t.table_comment`
 // We join with a subquery that materializes the data from `information_schema.innodb_sys_tablespaces`
 // early for performance reasons. This effectively causes only a single read of `information_schema.innodb_tablespaces`
 // per query.
+// Note the following:
+//   - We use UNION ALL to deal differently with partitioned tables vs. non-partitioned tables.
+//     Originally, the query handled both, but that introduced "WHERE ... OR" conditions that led to poor query
+//     optimization. By separating to UNION ALL we remove all "OR" conditions.
+//   - We utilize `INFORMATION_SCHEMA`.`TABLES`.`CREATE_OPTIONS` column to do early pruning before the JOIN.
+//   - `TABLES`.`TABLE_NAME` has `utf8mb4_0900_ai_ci` collation.  `INNODB_TABLESPACES`.`NAME` has `utf8mb3_general_ci`.
+//     We normalize the collation to get better query performance (we force the casting at the time of our choosing)
 const TablesWithSize80 = `SELECT t.table_name,
-	t.table_type,
-	UNIX_TIMESTAMP(t.create_time),
-	t.table_comment,
-	SUM(i.file_size),
-	SUM(i.allocated_size)
-FROM information_schema.tables t
-LEFT JOIN information_schema.innodb_tablespaces i
-	ON i.name LIKE CONCAT(database(), '/%') AND (i.name = CONCAT(t.table_schema, '/', t.table_name) OR i.name LIKE CONCAT(t.table_schema, '/', t.table_name, '#p#%'))
-WHERE t.table_schema = database()
-GROUP BY t.table_name, t.table_type, t.create_time, t.table_comment`
+		t.table_type,
+		UNIX_TIMESTAMP(t.create_time),
+		t.table_comment,
+		i.file_size,
+		i.allocated_size
+	FROM information_schema.tables t
+		LEFT JOIN information_schema.innodb_tablespaces i
+	ON i.name = CONCAT(t.table_schema, '/', t.table_name) COLLATE utf8_general_ci
```

Comment:
Hm. Shouldn't we collate to at least a `utf8mb4` collation, because that's a strict superset of `utf8mb3`, and thus all values valid in `TABLES.TABLE_NAME` can be handled?

Reply:
The table name has a `utf8mb3` character set:

```sql
> select * from information_schema.columns where table_name='TABLES' and column_name='TABLE_NAME'\G
*************************** 1. row ***************************
           TABLE_CATALOG: def
            TABLE_SCHEMA: information_schema
              TABLE_NAME: TABLES
             COLUMN_NAME: TABLE_NAME
        ORDINAL_POSITION: 3
          COLUMN_DEFAULT: NULL
             IS_NULLABLE: YES
               DATA_TYPE: varchar
CHARACTER_MAXIMUM_LENGTH: 64
  CHARACTER_OCTET_LENGTH: 192
       NUMERIC_PRECISION: NULL
           NUMERIC_SCALE: NULL
      DATETIME_PRECISION: NULL
      CHARACTER_SET_NAME: utf8mb3
          COLLATION_NAME: utf8mb3_tolower_ci
             COLUMN_TYPE: varchar(64)
```